crates/matrix-sdk-indexeddb/src/crypto_store/indexeddb_serializer.rs

```diff
@@ -0,0 +1,164 @@
+
+/// Handles the functionality of serializing and encrypting data for the
+/// indexeddb store.
+pub struct IndexeddbSerializer {
+    store_cipher: Option<Arc<StoreCipher>>,
+}
+
+impl IndexeddbSerializer {
+    pub fn new(store_cipher: Option<Arc<StoreCipher>>) -> Self {
+        Self { store_cipher }
+    }
+
+    /// Hash the given key securely for the given tablename, using the store
+    /// cipher.
+    ///
+    /// First calls [`SafeEncode::as_encoded_string`]
+    /// on the `key` to encode it into a formatted string.
+    ///
+    /// Then, if a cipher is configured, hashes the formatted key and returns
+    /// the hash encoded as unpadded base64.
+    ///
+    /// If no cipher is configured, just returns the formatted key.
+    ///
+    /// This is faster than [`serialize_value`] and reliably gives the same
+    /// output for the same input, making it suitable for index keys.
+    pub fn encode_key<T>(&self, table_name: &str, key: T) -> JsValue
+    where
+        T: SafeEncode,
+    {
+        self.encode_key_as_string(table_name, key).into()
+    }
+
+    /// Hash the given key securely for the given tablename, using the store
+    /// cipher.
+    ///
+    /// The same as [`encode_key`], but stops short of converting the resulting
+    /// base64 string into a JsValue
+    pub fn encode_key_as_string<T>(&self, table_name: &str, key: T) -> String
+    where
+        T: SafeEncode,
+    {
+        match &self.store_cipher {
+            Some(cipher) => key.as_secure_string(table_name, cipher),
+            None => key.as_encoded_string(),
+        }
+    }
+
+    pub fn encode_to_range<T>(
+        &self,
+        table_name: &str,
+        key: T,
+    ) -> Result<IdbKeyRange, IndexeddbCryptoStoreError>
+    where
+        T: SafeEncode,
+    {
+        match &self.store_cipher {
+            Some(cipher) => key.encode_to_range_secure(table_name, cipher),
+            None => key.encode_to_range(),
+        }
+        .map_err(|e| IndexeddbCryptoStoreError::DomException {
+            code: 0,
+            name: "IdbKeyRangeMakeError".to_owned(),
+            message: e,
+        })
+    }
+
+    /// Encode the value for storage as a value in indexeddb.
+    ///
+    /// First, serialise the given value as JSON.
+    ///
+    /// Then, if a store cipher is enabled, encrypt the JSON string using the
+    /// configured store cipher, giving a byte array. Then, wrap the byte
+    /// array as a `JsValue`.
+    ///
+    /// If no cipher is enabled, deserialises the JSON string again giving a JS
+    /// object.
+    pub fn serialize_value(&self, value: &impl Serialize) -> Result<JsValue, CryptoStoreError> {
+        if let Some(cipher) = &self.store_cipher {
+            let value = cipher.encrypt_value(value).map_err(CryptoStoreError::backend)?;
+
+            // Turn the Vec<u8> into a Javascript-side `Array<number>`.
+            // XXX Isn't there a way to do this that *doesn't* involve going via a JSON
+            // string?
```

Comment:
Sorry, there's at least comment changes and code motion in this commit, and since it's a much bigger chunk than the previous, I'll request to split it into a code motion one first, and then the changes to the code as a second commit please.

Reply:
Yup, sorry about this. The comments aren't particularly significant imho which is why I rolled them all together, but that was bad of me.

I've now split this into three separate commits.